### ISSUE TITLE
Adding S3 client

### ISSUE THREAD

**Issue by kubo6472** [girlbossceo/conduwuit#93]
No idea what the rewrite mentioned [here](https://gitlab.com/famedly/conduit/-/merge_requests/384) was about, but I wanted to ask if there are any plans at implementing an option to easily specify S3, B2, R2... bucket to store all the media in. Ideally with configurable options to replicate both locally and remotely, or to only save the files remotely.

On an unrelated note, specifically for Backblaze B2, there seems to be a pricing waive of eggress fees when using cloudflare workers to access the data, e.g. with something like [this](https://github.com/karldreher/b2-cloudflare-worker). Maybe this will result in an easier implementation, if it just loaded URLs for the media. However I doubt it has any security and not everyone uses Cloudflare.

> You can take advantage of Cloudflare's performance and the free data transfer between Backblaze B2 and Cloudflare: [ref.](https://www.backblaze.com/docs/cloud-storage-deliver-private-backblaze-b2-content-through-cloudflare-cdn)

**Comment by girlbossceo** [comment]
You are right it mostly appears to be just a bunch of config options. I can consider merging this into my fork, but I will only feel comfortable doing it if I add a clear and explicit warning that this (currently) has *no* cache and someone malicious can likely incur high-costs to you just by spamming media requests, and performance of media retrieval may be worse when using S3 due to high round trip times. And I will keep this issue open because the existing implementation requires these improvements, and likely can be improved with just supporting media redirection from MSC3860.

**Comment by girlbossceo** [comment]
Will merge https://gitlab.com/famedly/conduit/-/merge_requests/384 in with the warnings I want, and do a check on the code, but this issue will stay open with regards to what I said. Someone malicious can incur high costs to you by spamming requests (could be mitigated with https://github.com/girlbossceo/conduwuit/issues/98) and performance may or may not be significantly worse if roundtrip times are high with no cache.

**Comment by zoujiaqing** [comment]
Use minio instead of aws s3.

**Comment by AndSDev** [comment]
> Someone malicious can incur high costs to you by spamming requests

The cache will not save you from an attacker (will help, but won't save).
Matrix does not have protection against spam requests (except of rate-limits for requests).
So a warning is needed anyway.

P.S. This MR was designed for use with minio. Ideally, conduit should be a stateless service.
